src/libstore/s3-binary-cache-store.cc

```diff
@@ -241,8 +241,11 @@ struct S3BinaryCacheStoreImpl : public S3BinaryCacheStore
             auto & error = res.GetError();
             if (error.GetErrorType() == Aws::S3::S3Errors::RESOURCE_NOT_FOUND
                 || error.GetErrorType() == Aws::S3::S3Errors::NO_SUCH_KEY
+                // If bucket listing is disabled, 404s turn into 403s
+                || error.GetErrorType() == Aws::S3::S3Errors::ACCESS_DENIED
                 || (error.GetErrorType() == Aws::S3::S3Errors::UNKNOWN // FIXME
-                    && error.GetMessage().find("404") != std::string::npos))
+                    && (error.GetMessage().find("404") != std::string::npos
+                    || error.GetMessage().find("403") != std::string::npos)))
```

Comment:
Is checking for the 403 string necessary?

Reply:
Not really sure! I figured that you had some sort of weird edge case that led you to checking for `UNKNOWN` with 404, and had no confidence it wouldn't strike on a 403 too. I can remove it if you don't think it's necessary.